### PR TITLE
✨ Implement PV and PVC controllers

### PR DESCRIPTION
**Pull request by leseb**
## Summary

This implements two new syncers as laid out in
https://github.com/kcp-dev/kcp/issues/1981.

## Related issue(s)

Fixes: https://github.com/kcp-dev/kcp/issues/1981
Signed-off-by: Sébastien Han <seb@redhat.com>

### CODE REVIEW

**Comment by leseb** [comment]
I still haven't had the chance to test this code but it's ready for initial reviews. Thanks!

File: pkg/syncer/status/status_controller.go
Comment (davidfestal) [inline pkg/syncer/status/status_controller.go:101]: In order not to leave an event, we should modify ` deepEqualFinalizersAndStatus` in order to also compare the value of the value of the delaying annotation, so that the removal of the annotation would enqueue of the resource again.

File: pkg/syncer/status/status_process.go
Comment (davidfestal) [inline pkg/syncer/status/status_process.go:53]:
```suggestion
func hasDelayStatusSyncingAnnotation(unstrob *unstructured.Unstructured) bool {
```

File: pkg/syncer/storage/storage_pvc_controller.go
Comment (davidfestal) [inline pkg/syncer/storage/storage_pvc_controller.go:97]: I wouldn't do any retry here.
When there is an error, the main process function return the error and the resource is queued again.
cf https://github.com/kcp-dev/kcp/pull/2338/files#diff-59babfb03aea1b519e11dc21eaa7df1f4e810988bb0ccf4e96a6a0cd15bb26e5R177

File: pkg/syncer/storage/storage_pv_controller.go
Comment (davidfestal) [inline pkg/syncer/storage/storage_pv_controller.go:86]: I wouldn't do any retry here.
When there is an error, the main process function return the error and the resource is queued again.
cf https://github.com/kcp-dev/kcp/pull/2338/files#diff-59babfb03aea1b519e11dc21eaa7df1f4e810988bb0ccf4e96a6a0cd15bb26e5R177

File: pkg/syncer/storage/storage_pv_controller.go
Comment (davidfestal) [inline pkg/syncer/storage/storage_pv_controller.go:90]: You'd rather let the caller define his own message.

File: pkg/syncer/storage/storage_pv_controller.go
Comment (davidfestal) [inline pkg/syncer/storage/storage_pv_controller.go:109]: We should add a TODO here, so that we in fact point to the Upsyncer virtual workspace informer here (once @bipuladh  PR is available)

File: pkg/syncer/storage/storage_pv_process.go
Comment (davidfestal) [inline pkg/syncer/storage/storage_pv_process.go:58]: The namespace locator annotation is on downstream resources, to indicate the corresponding upstream information.

Comment (davidfestal) [reply pkg/syncer/storage/storage_pv_process.go:58]: So here you would just have to create a namespace locator containing the same information as what you downstream PV would contain. You should have all the information (upstream workspace, namespace, and synctarget info)

Comment (davidfestal) [reply pkg/syncer/storage/storage_pv_process.go:58]: Like this (found in SpecProcess):
```
	desiredNSLocator := shared.NewNamespaceLocator(clusterName, c.syncTargetWorkspace, c.syncTargetUID, c.syncTargetName, upstreamNamespace)
	jsonNSLocator, err := json.Marshal(desiredNSLocator)
	if err != nil {
		return err
	}

	var downstreamNamespace string
       downstreamNamespaces, err := c.downstreamNSInformer.Informer().GetIndexer().ByIndex(byNamespaceLocatorIndexName, string(jsonNSLocator))
```

Comment (davidfestal) [reply pkg/syncer/storage/storage_pv_process.go:58]: The PVC upstream namespace you would find in the upstream PV claimref.

File: pkg/syncer/storage/storage_pvc_controller.go
Comment (davidfestal) [inline pkg/syncer/storage/storage_pvc_controller.go:109]: Same as above

File: pkg/syncer/storage/storage_pvc_process.go
Comment (davidfestal) [inline pkg/syncer/storage/storage_pvc_process.go:54]: We have to take in account the NoFound case here, and according to the logic, possibly ignore the reconciliation without failing.

File: pkg/syncer/storage/storage_pvc_process.go
Comment (davidfestal) [inline pkg/syncer/storage/storage_pvc_process.go:66]: We cannot probably depend on generation here.
We have to think a bit more about how we st it here and unset it in the PV controller, so that both controllers wouldn't fight.

Comment (davidfestal) [reply pkg/syncer/storage/storage_pvc_process.go:66]: probably setting this whenever the PVC is not bound.

File: pkg/syncer/storage/storage_pvc_process.go
Comment (leseb) [reply pkg/syncer/storage/storage_pvc_process.go:54]: So this would handle the case where the object was removed during the reconcile loop?

Comment (davidfestal) [reply pkg/syncer/storage/storage_pvc_process.go:54]: yes, exactky.

File: pkg/syncer/shared/namespace.go
Comment (stevekuznetsov) [inline pkg/syncer/shared/namespace.go:111]: If this is exposed for other users, I'd expect a constant for the name of the index as well as a function to format the key as well

File: pkg/syncer/status/status_controller.go
Comment (stevekuznetsov) [inline pkg/syncer/status/status_controller.go:100]: Why does one function need to handle both cases?

File: pkg/syncer/status/status_controller.go
Comment (stevekuznetsov) [inline pkg/syncer/status/status_controller.go:89]: This kind of looks like a band-aid - why isn't it?

File: pkg/syncer/status/status_process_test.go
Comment (stevekuznetsov) [inline pkg/syncer/status/status_process_test.go:804]: nit: test names and logging output is best written for folks that accidentally break these but have no idea how the system works - names like `"true / true"` and the `t.Errorf()` output in this test are opaque and fairly arcane, let's consider how we might make them more clear.

File: pkg/syncer/storage/storage_pv_controller.go
Comment (stevekuznetsov) [inline pkg/syncer/storage/storage_pv_controller.go:112]: We should use the indices helpers here.

File: pkg/syncer/storage/storage_pv_controller.go
Comment (stevekuznetsov) [inline pkg/syncer/storage/storage_pv_controller.go:119]: Let's handle the cases where there are no volumes, or many - better to emit some admin-visible signal here.

File: pkg/syncer/storage/storage_pv_process.go
Comment (stevekuznetsov) [inline pkg/syncer/storage/storage_pv_process.go:96]: nit: logging key-value keys should not contain spaces

File: pkg/syncer/storage/storage_pvc_controller.go
Comment (stevekuznetsov) [inline pkg/syncer/storage/storage_pvc_controller.go:96]: nit: you can just return the values from the client call here, the `if` logic does not change anything

File: pkg/syncer/storage/storage_pv_controller.go
Comment (stevekuznetsov) [inline pkg/syncer/storage/storage_pv_controller.go:151]: If you are queueing objects from the upstream (kcp), you need the cluster-aware keyfunctions & splitters

File: pkg/syncer/storage/storage_pvc_process.go
Comment (stevekuznetsov) [inline pkg/syncer/storage/storage_pvc_process.go:124]: We have a patcher utility, let's use that

File: pkg/syncer/status/status_controller.go
Comment (leseb) [reply pkg/syncer/status/status_controller.go:89]: A band-aid?

File: pkg/syncer/status/status_process_test.go
Comment (leseb) [reply pkg/syncer/status/status_process_test.go:804]: Well it's a "simple" OR case so I didn't think too much 😅 , I'll try to think of something better.

File: hack/logcheck.out
Comment (leseb) [reply hack/logcheck.out:2]: I think I missed something during my last rebase, thanks for pointing it out.

File: pkg/syncer/status/status_controller.go
Comment (leseb) [reply pkg/syncer/status/status_controller.go:100]: This way it's easier to return true if the annotation got removed and the finalizer/status did not change. So testing everything at once felt clearer.

File: pkg/syncer/storage/storage_pvc_process.go
Comment (leseb) [reply pkg/syncer/storage/storage_pvc_process.go:124]: Can you point me to it?

Comment (stevekuznetsov) [reply pkg/syncer/storage/storage_pvc_process.go:124]: Sure thing! Here: https://github.com/kcp-dev/kcp/blob/main/pkg/reconciler/committer/committer.go

File: pkg/syncer/status/status_controller.go
Comment (stevekuznetsov) [reply pkg/syncer/status/status_controller.go:89]: Why are we delaying the sync? Why is it for some period of time, as opposed to until some event occurs?

File: pkg/syncer/storage/storage_pv_controller.go
Comment (stevekuznetsov) [reply pkg/syncer/storage/storage_pv_controller.go:112]: Under `pkg/indexers` - here: https://github.com/kcp-dev/kcp/blob/5916c3cd79ca30b3ed6e8a8983d10034f1a926e2/pkg/indexers/indexers.go#L74-L87

File: pkg/syncer/storage/storage_pv_process.go
Comment (leseb) [reply pkg/syncer/storage/storage_pv_process.go:96]: I see numerous instances in the code with loggers using spaces, I'm a bit confused by this rule. Or perhaps I misunderstood what you meant?

Comment (stevekuznetsov) [reply pkg/syncer/storage/storage_pv_process.go:96]: All of them should stop - it's bad UX for selecting things. Consider the difficulty of using `jq` to select on the key `"annotation from downstream"`

File: pkg/syncer/storage/storage_pvc_process.go
Comment (leseb) [reply pkg/syncer/storage/storage_pvc_process.go:124]: Confirmed offline that we need a committer that is not cluster-aware. So we can revisit this a followup PR.

File: pkg/syncer/storage/storage_pv_process.go
Comment (leseb) [reply pkg/syncer/storage/storage_pv_process.go:96]: It's hard to build meaningful human-readable logs then, but anyways, thanks for clarifying.

File: pkg/syncer/storage/storage_pvc_process.go
Comment (leseb) [reply pkg/syncer/storage/storage_pvc_process.go:124]: Actually, I went ahead and implemented something. I'll split into individual commits at the end of the review.

File: pkg/syncer/storage/storage_pv_controller.go
Comment (leseb) [reply pkg/syncer/storage/storage_pv_controller.go:151]: I assume you meant `kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc()` so I went ahead and used it. Let me know if I misunderstood. Thanks!

File: pkg/syncer/storage/storage_pv_process.go
Comment (stevekuznetsov) [reply pkg/syncer/storage/storage_pv_process.go]: Structured log _keys_ are not for humans. Expect those to be used to generate e.g. columns in tables, or for selection, etc. _Values_ can be whatever you'd like :)

File: pkg/syncer/storage/storage_pv_process.go
Comment (leseb) [reply pkg/syncer/storage/storage_pv_process.go:98]: I decided to split because the whole logic started to be quite long, also slitting up as smaller functions allow for better/smaller unit tests.

File: pkg/syncer/storage/storage_pvc_process.go
Comment (leseb) [reply pkg/syncer/storage/storage_pvc_process.go:209]: Ditto

File: pkg/syncer/storage/storage_pvc_process.go
Comment (leseb) [reply pkg/syncer/storage/storage_pvc_process.go:227]: ditto

File: pkg/syncer/storage/storage_pv_process.go
Comment (davidfestal) [reply pkg/syncer/storage/storage_pv_process.go:98]: I don't really see the value of unit tests on such a small amount of code honestly.
I prefer high-coverage unit tests of the controller `process` method.

File: pkg/reconciler/committer/committer.go
Comment (stevekuznetsov) [inline pkg/reconciler/committer/committer.go:78]: We use the term `Scoped` to talk about clients inside of one cluster

File: pkg/syncer/shared/namespace.go
Comment (stevekuznetsov) [inline pkg/syncer/shared/namespace.go:35]: We have pkg/indexers for these helpers, let's co-locate

File: pkg/syncer/shared/namespace.go
Comment (stevekuznetsov) [inline pkg/syncer/shared/namespace.go:133]: Hm, this is not very useful. I would expect this to take `*NamespaceLocator` or similar to make it useful to callers.

File: pkg/reconciler/committer/committer.go
Comment (leseb) [reply pkg/reconciler/committer/committer.go:52]: Is a separate commit accessible too? That's my goal once the reviews are done and the PR has a few approvals. Thanks.

Comment (leseb) [reply pkg/reconciler/committer/committer.go:52]: https://github.com/kcp-dev/kcp/pull/2458

File: test/e2e/virtual/syncer/storage_pv_test.go
Comment (davidfestal) [inline test/e2e/virtual/syncer/storage_pv_test.go:117]: Using `.RequestPath()` is wrong here. It's going to put a URL where there should be a logical cluster `Name` (the unique ID).
I think `workspaceName` should be a logicalCluster.Name, and you could also pass the `workspacePath` as an argument to test cases as well. But these are 2 distinct things.

File: test/e2e/virtual/syncer/storage_pv_test.go
Comment (davidfestal) [inline test/e2e/virtual/syncer/storage_pv_test.go:226]: Yuu will need both the workspacePath, and the workspaceName, as you did in the PVC e2e test: 
https://github.com/kcp-dev/kcp/pull/2338/files#diff-c0313ded30e0be3ab94b6a461c4e05751f9e6b192b40afd57d93017130588219R254-R255

File: test/e2e/virtual/syncer/storage_pv_test.go
Comment (davidfestal) [inline test/e2e/virtual/syncer/storage_pv_test.go:18]: Could we move those 2 tests in a new `/test/e2e/syncer/storage` package ?

They don't properly test the syncer **virtual workspace** feature, but new controllers added inside the Syncer itself.

**Comment by davidfestal** [comment]
I created PR https://github.com/leseb/kcp/pull/2 against your branch (since it seems I cannot push directly to your PR forked branch). This fixes the pv e2e test.

**Review by davidfestal — CHANGES_REQUESTED** [review]
Before merging this PR, we should add a feature flag to enable these actions: `KCPStorageMove`.

- See here for defining the new feature flag: https://github.com/leseb/kcp/blob/58ee6d48c9b5c3f4a69d076a954344785ac42505/pkg/features/kcp_features.go#L50

- See here for testing that feature flag presence in the syncer:  https://github.com/leseb/kcp/blob/58ee6d48c9b5c3f4a69d076a954344785ac42505/pkg/syncer/syncer.go#L424

- see here for setting the feature flag before a specific e2e test (which you will need): https://github.com/leseb/kcp/blob/58ee6d48c9b5c3f4a69d076a954344785ac42505/test/e2e/syncer/tunnels_test.go#L56

File: test/e2e/virtual/syncer/storage_pv_test.go
Comment (davidfestal) [inline test/e2e/virtual/syncer/storage_pv_test.go:127]: strange that you need to take the logical cluster and UID from the syncTarget, though you can get them from the SyncerConfig, as you did above in line [79](https://github.com/kcp-dev/kcp/pull/2338/files#diff-aeaf51b0a6e9e0efee0df9bfca87018758b76a8eae5eca1da7695c7ce2949c9dR79).

Would be better if the method wold be consistent.

File: pkg/syncer/syncer.go
Comment (davidfestal) [inline pkg/syncer/syncer.go:374]: let's start the upstreamUpsyncerControllerManager in any case, and not condition that to the feature gate.

**Comment by davidfestal** [comment]
Do you think you could rebase the PR on top of main (should not hurt I assume) ?
A number of flakes on e2d tests have been fixed, so that it would be easier to understand if the failing tests in this PR come from this PR changes, or from old flakes.

**Pull request closed without merging**
